docs/design-proposals/multi-tenancy/multi-tenancy-access-control.md

```diff
@@ -18,27 +18,54 @@ The key changes to each authenticator is that it needs to retrieve the tenant in
+
+### 2.Token File Authenticator
+With --token-auth-file=/path/to/tokenfile, token is read from tokenfile with the following format:
+
+```text
+[token],[user],[uid],[groups...],[other data]
+```
+
+Token file is prepared by system admin and supplied to cluster at start-up time. To support multi-tenancy, this format is changed to:
+
+```text
+[token],[user],[uid],[groups...],[other data],,[tenant]
+```
+
+The empty ",," at the end indicates a tenant name is to follow.  This new format provides backward compatibility. If no such field of arrangement exists, it will be identified as the old format and thus tenant is defaulted to "system".
+
+### 3.Basic Auth File Authenticator
+With--basic-auth-file=/path/to/somefile, token is read from tokenfile with the following format:
 
 ```text
-openssl req -new -key demo.pem -out demo-csr.pem -subj "/CN=demo/O=tenant1/OU=app1/OU=app2"
+[password],[user],[uid],[groups...],[other data]
 ```
 
-The certificate authenticator will map it to the following user:
-* The user name is "demo".
-* The user belongs to tenant "tenant1".
-* The user has a membership of group "app1" and "app1".
+Password file is prepared by system admin and supplied to cluster at start-up time. To support multi-tenancy, this format is changed to:
 
-### 2.Token Authenticator
-(**TBD**)
+```text
+[password],[user],[uid],[groups...],[other data],,[tenant]
+```
 
-### 3.Basic Authenticator
-(**TBD**)
+The empty ",," at the end indicates a tenant name is to follow.  This new format provides backward compatibility. If no such field of arrangement exists, it will be identified as the old format and thus tenant is defaulted to "system".
 
 ### 4.Service Account Authenticator
-(**TBD**)
+Service account can use JSON Web Token (JWT) as  a bearer token to authenticate with apiserver.  Both service account and sercret will be changed to tenant-scope. The tenant name is to be added to the bearer token so that it can be retrieved and verified during authentication.
+
+### 5.Webhook Token Authentication
+
+Arktos delegates user management to external components such as OpenStack Keystone through webhook token authentication uisng [k8s-keystone-auth](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-keystone-webhook-authenticator-and-authorizer.md). A token is issued by Keystone once a user successfully logs in at Keystone. This token is passed from kubectl to apiserver for authentication as a bearer token, and in turn it is passed by k8s-keystone-auth to the Keystone service for validation. Upon success, user info such as user's name, domain, project, roles is returned to Arktos. For arktos, Keystone domain is mapped to "tenant", while the combination of Keystone project and role maps to Arktos role, and Keystone role assignment which contains project, role and user is naturally mapped to the Arktos rolebinding. Apiserver will be changed to extract the domain returned from Keystone and set it to tenant in context.
```

Comment:
Need to give conclusion here that if we need code changes for Webhook Token Auth.

Reply:
reworded.